lib/hexdocs/search/typesense.ex

```diff
@@ -0,0 +1,76 @@
+defmodule Hexdocs.Search.Typesense do
+  require Logger
+
+  @behaviour Hexdocs.Search
+
+  @collection "hexdocs"
+
+  @impl true
+  def index(package, version, search_items) do
+    full_package = "#{package}-#{version}"
+
+    ndjson =
+      Enum.map(search_items, fn item ->
+        json =
+          Map.take(item, ["type", "ref", "title", "doc"])
+          |> Map.put("package", full_package)
+          |> Jason.encode_to_iodata!()
+
+        [json, ?\n]
+      end)
+
+    url = url("collections/#{@collection}/documents/import?action=create")
+    headers = headers([{"content-type", "text/plain"}])
+
+    case :hackney.post(url, headers, ndjson) do
+      {:ok, 200, _resp_headers, ref} ->
+        process_results(package, version, ref)
+        :ok
+
+      {:ok, status, _resp_headers, _ref} ->
+        Logger.error("Failed to index search items for #{package} #{version}: status=#{status}")
+
+      {:error, reason} ->
+        Logger.error("Failed to index search items #{package} #{version}: #{inspect(reason)}")
+    end
+  end
+
+  @spec process_results(String.t(), Version.t(), :hackney.client_ref()) :: :ok
+  defp process_results(package, version, ref) do
+    case :hackney.stream_body(ref) do
+      {:ok, ndjson} ->
+        ndjson
+        |> String.splitter("\n")
+        |> Stream.each(fn json ->
+          case Jason.decode!(json) do
```

Comment:
dunno if it'd make a difference performance wise but we should be able to use the new :json module for this:

```elixir
iex> binary = ~s({"x":1}\n{"x":2}\n{"x":3})
iex> acc = nil
iex> {decoded, acc, binary} = :json.decode(binary, acc, %{})
{%{"x" => 1}, nil, "{\"x\":2}\n{\"x\":3}\n"}
iex> {decoded, acc, binary} = :json.decode(binary, acc, %{})
{%{"x" => 2}, nil, "{\"x\":3}\n"}
iex> {decoded, acc, binary} = :json.decode(binary, acc, %{})
{%{"x" => 3}, nil, ""}
```

We can totally ship this as is, I am just mentioning this in case it piques your interest. Not gonna lie, I'd love :json-based _ndjson_ eager and streaming decoder for Req. ;)

Reply:
Hexdocs is using Erlang 25 right now.

Reply:
Yeah, that'd be great and appreciated, no rush though!

Reply:
> no rush though!

Updating versions is my favorite activity for some reason: https://github.com/hexpm/hexdocs/pull/45 :)